src/serverobjecttype.cpp

```diff
@@ -84,25 +87,204 @@ NodeId ServerObjectType::append(
 
 //*****************************************************************************
 
-NodeId ServerObjectType::addInstance(
-    const std::string&  name,
-    const NodeId&       parent,
-    const NodeId&       requestNodeId   /*= NodeId::Null*/,
-    NodeContext*        context         /*= nullptr*/) {
-
-    NodeId newNode;
-    newNode.notNull();
-
+NodeId ServerObjectType::addInstance(const std::string& name,
+                                     const NodeId& parent,
+                                     NodeId& nodeId,
+                                     const NodeId& requestNodeId,
+                                     NodeContext* context)
+{
     bool ret = m_server.addInstance(
         name,
         requestNodeId,
         parent,
         m_typeId,
-        newNode,
+        nodeId,
         context);
 
    UAPRINTLASTERROR(m_server.lastError());
-   return ret ? newNode : NodeId::Null;
+    return ret ? nodeId : NodeId::Null;
+}
+
+bool ServerObjectType::addObjectTypeFolder(const std::string& childName,
+                         const NodeId& parent,
+                         NodeId& nodeId,
+                         NodeId& requestNodeId,
+                         bool mandatory)
+{
+    NodeId newNode;
+    newNode.notNull();
+
+    if (m_server.addFolder(parent, childName, newNode, requestNodeId)) {
+        if (mandatory) {
+            return m_server.addReference(newNode,
+                                         NodeId::HasModellingRule,
+                                         ExpandedNodeId::ModellingRuleMandatory,
+                                         true);
+        }
+        if (!nodeId.isNull())
+            nodeId = newNode;
+        return true;
+    }
+    return false;
+}
+
+bool ServerObjectType::setMandatory(const NodeId& n1)
+{
+    return m_server.addReference(n1, NodeId::HasModellingRule, ExpandedNodeId::ModellingRuleMandatory, true) ==
+           UA_STATUSCODE_GOOD;
 }
 
+/*!
+    \brief addObjectTypeVariable
+    \param n
+    \param parent
+    \param nodeiD
+    \param mandatory
+    \return
+*/
+template <typename T>
+bool ServerObjectType::addObjectTypeVariable(
```

Comment:
I don't think moving the definition of a templated function into a cpp file is a good idea.
cf: [https://raymii.org/s/snippets/Cpp_template_definitions_in_a_cpp_file_instead_of_header.html](https://raymii.org/s/snippets/Cpp_template_definitions_in_a_cpp_file_instead_of_header.html) and
[https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file](https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file)